dev-clean/src/main/scala/sai/llsc/EngineBase.scala

```diff
@@ -193,21 +191,21 @@ trait EngineBase extends SAIOps { self: BasicDefs with ValueDefs =>
     val (size, align) = getTySizeAlign(real_ty)
```

Comment:
Unaware of your current plan, there seems to be much inconsistency in this commit as far as I'm concerned. So maybe I'd provide some rationale here.

`getTySizeAlign` is a useful utility, but it should not be considered as lightweight as `getTySize` is, especially when taking structs into sight. So one key goal of the initial version of `evalHeapConstWithAlign` is to run this structure-padding calculation only once. If not so we can simply call `getTySizeAlign` wherever we need the alignment information, and revert `evalHeapConstWithAlign` to `evalHeapConst`. Either can do, but we don't have to keep them both.

Reply:
Thanks for weighing in! I was just about to ask for your review today. 

Yes, I'm aware that the refactored `evalHeapConstWithAlign` is a bit fishy. My main goal in this refactor was to allow a value to generate its own raw bytes sequence so that we can implement the "baseline" memory model.

> `getTySizeAlign` is a useful utility, but it should not be considered as lightweight as `getTySize` is, especially when taking structs into sight.

But `getTySize` is implemented by calling `getTySizeAlign`, they should be the same "lightweight"?

> So one key goal of the initial version of evalHeapConstWithAlign is to run this structure-padding calculation only once. If not so we can simply call getTySizeAlign wherever we need the alignment information, and revert evalHeapConstWithAlign to evalHeapConst. Either can do, but we don't have to keep them both.

Hmm, say in the previous version, we recursively call `evalHeapConstWithAlign` with `StructCalc.concat` on a struct const, then the recursion would call `getTySizeAlign` on each field of the struct, is that right? Then what exactly runs once?

Ideally, I think `evalHeap*` could be just a simple structurally recursive function over `Constant`, and I would prefer `align` could be somewhat less intrusive. I'm also now more inclined that `healEval*` functions should not be confused with engine-level `eval` function, since the former operates on the level of our memory model of choice (i.e. mostly bytes), but the latter produces whole/logical values.

The other option I can see is closer to the previous version that we divide syntax `Constant` into several sub-categories, for example, atomic constant (int, flat, etc) and complex constant (like struct/array). And we have non-overlap functions to handle them. This is similar to the common practice when dealing with A-normal form languages.

> Minor: it does not make much sense to me to return a pointer for the PtrToIntExpr. 

Yes, I should extract the integer from it.

Reply:
> The other option I can see is closer to the previous version that we divide syntax Constant into several sub-categories, for example, atomic constant (int, float, etc) and complex constant (like struct/array). And we have non-overlap functions to handle them. This is similar to the common practice when dealing with A-normal form languages.

@Tantalus13A98B5F to make this point clear, I just did a quick version of it (https://github.com/Kraks/sai/pull/52/commits/dded1cfd2173d492c446b7bd1edcf24dca848731). Seems to be more clear?

Reply:
I'm finding myself not so sure about the `evalHeapConst` things. Such complex folds are really mind-bending for me orz.

For `getTySize`, there's an early version without alignment, but it also did recursive calls on structs (which wasn't on my mind when I'm writing "lightweight"). In `getTySizeAlign` there is nothing much added: fields are still visited just once each, with `padding` called on every pair of consecutive fields.

In `evalHeapConst` without alignment consideration, we need `getTySize` for only the atomic types, which are the fast paths. Things get a little bit trickier for `evalHeapConstWithAlign`, in that we start to need alignment info for every field, compound or atomic. If we don't carry alignment info with `evalHeapConst`, then for every added level of nested struct, we need an added traversal by calling `getTySizeAlign`. But again, this only affects nested structs, which should be rare.

To conclude, `evalHeapConstWithAlign` works by visiting each atomic field only once (by recursive call on the struct case), and `getTySizeAlign` is called on only atomic fields. Such "traversing only once" scheme seems elegant to me in the first place, but twice is not really worse, and like you said we can forget about the `align` in `evalHeapConst`, which is also nice.

Reply:
> The other option I can see is closer to the previous version that we divide syntax Constant into several sub-categories, for example, atomic constant (int, float, etc) and complex constant (like struct/array). And we have non-overlap functions to handle them. This is similar to the common practice when dealing with A-normal form languages.

As a daydreamer I would expect something like: in a `match` we return different types of values for different cases; and then we do an extra `match` to lift up the possible simple values. As this does not make sense in Scala (can we?) I'm fine with the function approach, but personally I won't make them top-level.

Reply:
Thanks for the explanation. I think now `evalHeapConstWithAlign` has the same traverse behavior as previous, but just the code is organized a bit differently. I wouldn't mind further refactoring it for proper access control or encapsulation, but let's see if it passes test cases first.